File: frontend/src/api/sql_studio.ts
import { useMutation, useQuery, useQueryClient } from "@tanstack/react-query";
import { apiRequest } from "./client";

// ---------------------------------------------------------------------------
// 3.2.4 dry-run validate
// ---------------------------------------------------------------------------
export interface SqlValidateResponse {
  valid: boolean;
  error: string | null;
  referenced_tables: string[];
}

export function useValidateSql() {
  return useMutation({
    mutationFn: (sql: string) =>
      apiRequest<SqlValidateResponse>("/v1/sql-studio/validate", {
        method: "POST",
        body: { sql },
      }),
  });
}

// ---------------------------------------------------------------------------
// 3.2.5 preview / explain
// ---------------------------------------------------------------------------
export interface SqlPreviewRequest {
  sql: string;
  limit?: number;
  sql_query_version_id?: number | null;
}

export interface SqlPreviewResponse {
  columns: string[];
  rows: unknown[][];
  row_count: number;
  truncated: boolean;
  elapsed_ms: number;
}

export function usePreviewSql() {
  return useMutation({
    mutationFn: (body: SqlPreviewRequest) =>
      apiRequest<SqlPreviewResponse>("/v1/sql-studio/preview", {
        method: "POST",
        body,
      }),
  });
}

export interface SqlExplainResponse {
  plan_json: Record<string, unknown>[];
  elapsed_ms: number;
}

export function useExplainSql() {
  return useMutation({
    mutationFn: (sql: string) =>
      apiRequest<SqlExplainResponse>("/v1/sql-studio/explain", {
        method: "POST",
        body: { sql },
      }),
  });
}

// ---------------------------------------------------------------------------
// 3.2.5 SqlQuery / SqlQueryVersion CRUD
// ---------------------------------------------------------------------------
export type SqlVersionStatus =
  | "DRAFT"
  | "PENDING"
  | "APPROVED"
  | "REJECTED"
  | "SUPERSEDED";

export interface SqlQueryVersionOut {
  sql_query_version_id: number;
  sql_query_id: number;
  version_no: number;
  sql_text: string;
  referenced_tables: string[];
  status: SqlVersionStatus;
  parent_version_id: number | null;
  submitted_by: number | null;
  submitted_at: string | null;
  reviewed_by: number | null;
  reviewed_at: string | null;
  review_comment: string | null;
  created_at: string;
}

export interface SqlQueryOut {
  sql_query_id: number;
  name: string;
  description: string | null;
  owner_user_id: number;
  current_version_id: number | null;
  created_at: string;
  updated_at: string;
}

export interface SqlQueryDetail extends SqlQueryOut {
  versions: SqlQueryVersionOut[];
}

export function useSqlQueries() {
  return useQuery({
    queryKey: ["sql-studio", "queries"],
    queryFn: () =>
      apiRequest<SqlQueryOut[]>("/v1/sql-studio/queries", {
        params: { limit: 200 },
      }),
  });
}

export function useSqlQueryDetail(queryId: number | null) {
  return useQuery({
    queryKey: ["sql-studio", "queries", queryId],
    enabled: queryId != null,
    queryFn: () =>
      apiRequest<SqlQueryDetail>(`/v1/sql-studio/queries/${queryId}`),
  });
}

export function useCreateSqlQuery() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (body: {
      name: string;
      description?: string | null;
      sql_text: string;
    }) =>
      apiRequest<SqlQueryDetail>("/v1/sql-studio/queries", {
        method: "POST",
        body,
      }),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["sql-studio", "queries"] });
    },
  });
}

export function useAddSqlVersion() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      queryId,
      sql_text,
    }: {
      queryId: number;
      sql_text: string;
    }) =>
      apiRequest<SqlQueryVersionOut>(
        `/v1/sql-studio/queries/${queryId}/versions`,
        { method: "POST", body: { sql_text } },
      ),
    onSuccess: (_, vars) => {
      qc.invalidateQueries({ queryKey: ["sql-studio", "queries", vars.queryId] });
    },
  });
}

export function useSubmitSqlVersion() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: (versionId: number) =>
      apiRequest<SqlQueryVersionOut>(
        `/v1/sql-studio/versions/${versionId}/submit`,
        { method: "POST" },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["sql-studio", "queries"] });
    },
  });
}

export function useApproveSqlVersion() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      versionId,
      comment,
    }: {
      versionId: number;
      comment?: string | null;
    }) =>
      apiRequest<SqlQueryVersionOut>(
        `/v1/sql-studio/versions/${versionId}/approve`,
        { method: "POST", body: { comment: comment ?? null } },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["sql-studio", "queries"] });
    },
  });
}

export function useRejectSqlVersion() {
  const qc = useQueryClient();
  return useMutation({
    mutationFn: ({
      versionId,
      comment,
    }: {
      versionId: number;
      comment?: string | null;
    }) =>
      apiRequest<SqlQueryVersionOut>(
        `/v1/sql-studio/versions/${versionId}/reject`,
        { method: "POST", body: { comment: comment ?? null } },
      ),
    onSuccess: () => {
      qc.invalidateQueries({ queryKey: ["sql-studio", "queries"] });
    },
  });
}

File: backend/app/schemas/sql_studio.py
"""SQL Studio API schemas (Phase 3.2.4 + 3.2.5)."""

from __future__ import annotations

from datetime import datetime
from typing import Any

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# 3.2.4 dry-run validate
# ---------------------------------------------------------------------------
class SqlValidateRequest(BaseModel):
    sql: str = Field(min_length=1, max_length=10_000)


class SqlValidateResponse(BaseModel):
    valid: bool
    error: str | None = None
    referenced_tables: list[str] = Field(default_factory=list)


# ---------------------------------------------------------------------------
# 3.2.5 preview / explain
# ---------------------------------------------------------------------------
class SqlPreviewRequest(BaseModel):
    sql: str = Field(min_length=1, max_length=20_000)
    limit: int = Field(default=1000, ge=1, le=10_000)
    sql_query_version_id: int | None = None


class SqlPreviewResponse(BaseModel):
    columns: list[str]
    rows: list[list[Any]]
    row_count: int
    truncated: bool
    elapsed_ms: int


class SqlExplainRequest(BaseModel):
    sql: str = Field(min_length=1, max_length=20_000)
    sql_query_version_id: int | None = None


class SqlExplainResponse(BaseModel):
    plan_json: list[dict[str, Any]]
    elapsed_ms: int


# ---------------------------------------------------------------------------
# 3.2.5 SqlQuery / SqlQueryVersion CRUD
# ---------------------------------------------------------------------------
class SqlQueryCreate(BaseModel):
    name: str = Field(min_length=1, max_length=200)
    description: str | None = Field(default=None, max_length=2_000)
    sql_text: str = Field(min_length=1, max_length=20_000)


class SqlVersionCreate(BaseModel):
    sql_text: str = Field(min_length=1, max_length=20_000)


class SqlVersionReview(BaseModel):
    """approve/reject 공용 — 코멘트만."""

    comment: str | None = Field(default=None, max_length=2_000)


class SqlQueryVersionOut(BaseModel):
    sql_query_version_id: int
    sql_query_id: int
    version_no: int
    sql_text: str
    referenced_tables: list[str]
    status: str
    parent_version_id: int | None
    submitted_by: int | None
    submitted_at: datetime | None
    reviewed_by: int | None
    reviewed_at: datetime | None
    review_comment: str | None
    created_at: datetime

    class Config:
        from_attributes = True


class SqlQueryOut(BaseModel):
    sql_query_id: int
    name: str
    description: str | None
    owner_user_id: int
    current_version_id: int | None
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class SqlQueryDetail(SqlQueryOut):
    versions: list[SqlQueryVersionOut]
